File: sdk/rust/src/planner.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PlanNode {
    pub id: String,
    pub kind: String,
    pub label: String,
    pub depends_on: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Plan {
    pub plan_id: String,
    pub goal: String,
    pub nodes: Vec<PlanNode>,
    pub metadata: HashMap<String, serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ImprovementProposal {
    pub proposal_id: String,
    pub lesson_id: String,
    pub target_node_id: Option<String>,
    pub action: String,
    pub detail: String,
    pub confidence: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Lesson {
    pub lesson_id: String,
    pub run_id: String,
    pub insight: String,
    pub severity: String,
    pub tags: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RankedPlan {
    pub plan: Plan,
    pub score: PlanScore,
    pub rank: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PlanScore {
    pub node_count: usize,
    pub depth: usize,
    pub risk: f64,
    pub confidence: String,
    pub complexity: f64,
    pub composite: f64,
}

pub struct GoalDecomposer;

impl GoalDecomposer {
    pub fn decompose(&self, goal: &str, constraints: Option<HashMap<String, serde_json::Value>>) -> Option<Plan> {
        let goal = goal.trim();
        if goal.is_empty() {
            return None;
        }
        let plan_id = sanitize_goal(goal);
        let steps = extract_verbs(goal);
        let mut nodes = Vec::new();
        for (i, step) in steps.iter().enumerate() {
            let deps = if i > 0 { vec![format!("step-{}", i)] } else { Vec::new() };
            nodes.push(PlanNode {
                id: format!("step-{}", i + 1),
                kind: "task".to_string(),
                label: step.clone(),
                depends_on: deps,
            });
        }
        let mut metadata = HashMap::new();
        metadata.insert("constraints".to_string(), serde_json::json!(constraints.unwrap_or_default()));
        Some(Plan {
            plan_id,
            goal: goal.to_string(),
            nodes,
            metadata,
        })
    }

    pub fn to_workflow(&self, plan: &Plan) -> Plan {
        plan.clone()
    }
}

pub struct Planner;

impl Planner {
    pub fn rank(&self, plans: &[Plan]) -> Vec<RankedPlan> {
        let mut scored: Vec<RankedPlan> = plans
            .iter()
            .map(|plan| RankedPlan {
                plan: plan.clone(),
                score: self.score(plan),
                rank: 0,
            })
            .collect();
        scored.sort_by(|a, b| b.score.composite.partial_cmp(&a.score.composite).unwrap());
        for (i, entry) in scored.iter_mut().enumerate() {
            entry.rank = i + 1;
        }
        scored
    }

    pub fn score(&self, plan: &Plan) -> PlanScore {
        let node_count = plan.nodes.len();
        let depth = max_depth(plan);
        let risk = 0.5;
        let confidence = "low";
        let complexity = node_count as f64 * 0.1 + depth as f64 * 0.2;
        let composite = (1.0 - risk - complexity * 0.1).max(0.0);
        PlanScore {
            node_count,
            depth,
            risk,
            confidence: confidence.to_string(),
            complexity: (complexity * 10000.0).round() / 10000.0,
            composite: (composite * 10000.0).round() / 10000.0,
        }
    }
}

pub struct Reflector {
    pub events: Vec<HashMap<String, serde_json::Value>>,
}

impl Reflector {
    pub fn new(events: Option<Vec<HashMap<String, serde_json::Value>>>) -> Self {
        Self { events: events.unwrap_or_default() }
    }

    pub fn reflect(&self, run_id: &str) -> Vec<Lesson> {
        let mut lessons = Vec::new();
        lessons.extend(self.detect_failure_patterns(run_id));
        lessons.extend(self.detect_inefficiencies(run_id));
        lessons.extend(self.detect_handoff_patterns(run_id));
        lessons
    }

    fn detect_failure_patterns(&self, run_id: &str) -> Vec<Lesson> {
        let mut lessons = Vec::new();
        let mut task_failures: HashMap<String, usize> = HashMap::new();
        for e in &self.events {
            if e.get("type").and_then(|v| v.as_str()) == Some("task_status") && e.get("status").and_then(|v| v.as_str()) == Some("[!]") {
                if let Some(tid) = e.get("task_id").and_then(|v| v.as_str()) {
                    *task_failures.entry(tid.to_string()).or_insert(0) += 1;
                }
            }
        }
        for (tid, count) in task_failures {
            if count >= 2 {
                lessons.push(Lesson {
                    lesson_id: format!("lesson-{}", lessons.len() + 1),
                    run_id: run_id.to_string(),
                    insight: format!("Task '{}' failed {} times; consider retry or fallback strategy.", tid, count),
                    severity: "warn".to_string(),
                    tags: vec!["failure".to_string(), tid],
                });
            }
        }
        lessons
    }

    fn detect_inefficiencies(&self, run_id: &str) -> Vec<Lesson> {
        let mut lessons = Vec::new();
        let mut claim_counts: HashMap<String, usize> = HashMap::new();
        for e in &self.events {
            if e.get("type").and_then(|v| v.as_str()) == Some("task_claim") {
                if let Some(tid) = e.get("task_id").and_then(|v| v.as_str()) {
                    *claim_counts.entry(tid.to_string()).or_insert(0) += 1;
                }
            }
        }
        for (tid, count) in claim_counts {
            if count >= 3 {
                lessons.push(Lesson {
                    lesson_id: format!("lesson-{}", lessons.len() + 1),
                    run_id: run_id.to_string(),
                    insight: format!("Task '{}' was claimed {} times; review ownership logic.", tid, count),
                    severity: "info".to_string(),
                    tags: vec!["efficiency".to_string(), tid],
                });
            }
        }
        lessons
    }

    fn detect_handoff_patterns(&self, run_id: &str) -> Vec<Lesson> {
        let mut handoffs = 0;
        for e in &self.events {
            if e.get("type").and_then(|v| v.as_str()) == Some("human_handoff") || e.get("status").and_then(|v| v.as_str()) == Some("[?]") {
                handoffs += 1;
            }
        }
        if handoffs > 1 {
            return vec![Lesson {
                lesson_id: "lesson-1".to_string(),
                run_id: run_id.to_string(),
                insight: format!("Run had {} human handoffs; consider automating or simplifying decision gates.", handoffs),
                severity: "warn".to_string(),
                tags: vec!["handoff".to_string()],
            }];
        }
        Vec::new()
    }

    pub fn improve_plan(&self, plan: &Plan, lessons: &[Lesson], constraints: Option<&HashMap<String, serde_json::Value>>) -> HashMap<String, serde_json::Value> {
        let mut nodes = plan.nodes.clone();
        let mut seen = std::collections::HashSet::new();
        let mut proposals = Vec::new();
        for lesson in lessons {
            if lesson.tags.contains(&"failure".to_string()) && lesson.tags.contains(&"failed".to_string()) {
                let target = extract_task_id(&lesson.insight);
                proposals.push(ImprovementProposal {
                    proposal_id: format!("prop-{}", proposals.len() + 1),
                    lesson_id: lesson.lesson_id.clone(),
                    target_node_id: Some(target),
                    action: "add_dependency".to_string(),
                    detail: format!("Add fallback or retry dependency for '{}' due to repeated failures.", target),
                    confidence: 0.75,
                });
            }
            if lesson.tags.contains(&"efficiency".to_string()) && lesson.tags.contains(&"claimed".to_string()) {
                let target = extract_task_id(&lesson.insight);
                proposals.push(ImprovementProposal {
                    proposal_id: format!("prop-{}", proposals.len() + 1),
                    lesson_id: lesson.lesson_id.clone(),
                    target_node_id: Some(target),
                    action: "reassign".to_string(),
                    detail: format!("Reassign '{}' to a more stable owner.", target),
                    confidence: 0.6,
                });
            }
            if lesson.tags.contains(&"handoff".to_string()) {
                proposals.push(ImprovementProposal {
                    proposal_id: format!("prop-{}", proposals.len() + 1),
                    lesson_id: lesson.lesson_id.clone(),
                    target_node_id: None,
                    action: "add_node".to_string(),
                    detail: "Add automation gate to reduce human handoff frequency.".to_string(),
                    confidence: 0.5,
                });
            }
        }
        let max_nodes = constraints.and_then(|c| c.get("max_nodes")).and_then(|v| v.as_u64()).map(|v| v as usize);
        for p in &proposals {
            if p.action == "add_node" && !seen.contains(&p.proposal_id) {
                if let Some(max) = max_nodes {
                    if nodes.len() >= max {
                        continue;
                    }
                }
                nodes.push(PlanNode {
                    id: format!("node-{}", p.proposal_id),
                    kind: "task".to_string(),
                    label: p.detail.clone(),
                    depends_on: Vec::new(),
                });
                seen.insert(p.proposal_id.clone());
            }
        }
        let mut improved = Plan {
            plan_id: plan.plan_id.clone(),
            goal: plan.goal.clone(),
            nodes,
            metadata: plan.metadata.clone(),
        };
        improved.metadata.insert("improvements".to_string(), serde_json::json!(proposals.iter().map(|p| p.action.clone()).collect::<Vec<_>>()));
        let mut result = HashMap::new();
        result.insert("plan".to_string(), serde_json::json!(improved));
        result.insert("proposals".to_string(), serde_json::json!(proposals));
        result
    }
}

pub struct CollabPlanner {
    pub estimator: Option<serde_json::Value>,
}

impl CollabPlanner {
    pub fn new(estimator: Option<serde_json::Value>) -> Self {
        Self { estimator }
    }

    pub fn build(&self, goal: &str, constraints: Option<HashMap<String, serde_json::Value>>) -> Option<Plan> {
        let constraints = constraints.unwrap_or_default();
        let decomposer = GoalDecomposer;
        let mut plan = decomposer.decompose(goal, Some(constraints.clone()))?;
        let planner = Planner;
        let ranked = planner.rank(&[plan.clone()]);
        if let Some(top) = ranked.first() {
            plan = top.plan.clone();
        }
        if self.estimator.is_some() {
            plan.metadata.insert("negotiation".to_string(), serde_json::json!("accepted"));
        }
        Some(plan)
    }
}

fn max_depth(plan: &Plan) -> usize {
    if plan.nodes.is_empty() {
        return 0;
    }
    let mut depths: HashMap<String, usize> = HashMap::new();
    for n in &plan.nodes {
        depths.insert(n.id.clone(), 1);
    }
    for n in &plan.nodes {
        for dep in &n.depends_on {
            if let Some(&d) = depths.get(dep) {
                depths.insert(n.id.clone(), depths[&n.id].max(d + 1));
            }
        }
    }
    *depths.values().max().unwrap_or(&0)
}

fn extract_task_id(insight: &str) -> String {
    let parts: Vec<&str> = insight.split('\'').collect();
    if parts.len() >= 2 {
        parts[1].to_string()
    } else {
        "unknown".to_string()
    }
}

fn sanitize_goal(goal: &str) -> String {
    let mut s = goal.trim().to_lowercase();
    s = s.replace(|c: char| !c.is_alphanumeric() && c != '_' && c != '-', "-");
    if s.len() > 40 {
        s.truncate(40);
    }
    if s.is_empty() {
        "plan".to_string()
    } else {
        s
    }
}

fn extract_verbs(goal: &str) -> Vec<String> {
    let words: Vec<&str> = goal.split_whitespace().collect();
    let mut verbs = Vec::new();
    for w in words {
        let clean = w.trim_matches(|c: char| ",.!?:;".contains(c));
        if !clean.is_empty() && clean.chars().next().map(|c| c.is_uppercase()).unwrap_or(false) {
            verbs.push(clean.to_string());
        }
    }
    if verbs.is_empty() {
        vec![goal.to_string()]
    } else {
        verbs
    }
}

File: sdk/go/planner.go
package alpgo

import (
	"fmt"
	"math"
	"sort"
	"strings"
)

type PlanNode struct {
	ID        string   `json:"id"`
	Kind      string   `json:"kind"`
	Label     string   `json:"label"`
	DependsOn []string `json:"depends_on"`
}

type Plan struct {
	PlanID    string              `json:"plan_id"`
	Goal      string              `json:"goal"`
	Nodes     []*PlanNode         `json:"nodes"`
	Metadata  map[string]any      `json:"metadata"`
}

type ImprovementProposal struct {
	ProposalID   string  `json:"proposal_id"`
	LessonID     string  `json:"lesson_id"`
	TargetNodeID string  `json:"target_node_id,omitempty"`
	Action       string  `json:"action"`
	Detail       string  `json:"detail"`
	Confidence   float64 `json:"confidence"`
}

type Lesson struct {
	LessonID string   `json:"lesson_id"`
	RunID    string   `json:"run_id"`
	Insight  string   `json:"insight"`
	Severity string   `json:"severity"`
	Tags     []string `json:"tags"`
}

func NewPlanNode(id, kind, label string, dependsOn []string) *PlanNode {
	return &PlanNode{
		ID:        id,
		Kind:      kind,
		Label:     label,
		DependsOn: dependsOn,
	}
}

func NewPlan(planID, goal string, nodes []*PlanNode, metadata map[string]any) *Plan {
	if metadata == nil {
		metadata = make(map[string]any)
	}
	return &Plan{
		PlanID:   planID,
		Goal:     goal,
		Nodes:    nodes,
		Metadata: metadata,
	}
}

type GoalDecomposer struct{}

func (g *GoalDecomposer) Decompose(goal string, constraints map[string]any) *Plan {
	goal = sanitizeGoal(goal)
	if goal == "" {
		return nil
	}
	planID := limitID(goal, 40)
	steps := extractVerbs(goal)
	nodes := make([]*PlanNode, 0, len(steps))
	for i, step := range steps {
		var deps []string
		if i > 0 {
			deps = []string{fmt.Sprintf("step-%d", i)}
		}
		nodes = append(nodes, NewPlanNode(fmt.Sprintf("step-%d", i+1), "task", step, deps))
	}
	meta := map[string]any{"constraints": constraints}
	if constraints == nil {
		meta["constraints"] = map[string]any{}
	}
	return NewPlan(planID, goal, nodes, meta)
}

func (g *GoalDecomposer) ToWorkflow(plan *Plan) *Plan {
	return plan
}

type Planner struct{}

func (p *Planner) Rank(plans []*Plan) []*RankedPlan {
	scored := make([]*RankedPlan, 0, len(plans))
	for _, plan := range plans {
		scored = append(scored, &RankedPlan{
			Plan:  plan,
			Score: p.Score(plan),
		})
	}
	sort.Slice(scored, func(i, j int) bool {
		return scored[i].Score.Composite > scored[j].Score.Composite
	})
	for i, entry := range scored {
		entry.Rank = i + 1
	}
	return scored
}

type RankedPlan struct {
	Plan  *Plan     `json:"plan"`
	Score PlanScore `json:"score"`
	Rank  int       `json:"rank"`
}

type PlanScore struct {
	NodeCount int     `json:"node_count"`
	Depth     int     `json:"depth"`
	Risk      float64 `json:"risk"`
	Confidence string `json:"confidence"`
	Complexity float64 `json:"complexity"`
	Composite float64 `json:"composite"`
}

func (p *Planner) Score(plan *Plan) PlanScore {
	nodeCount := len(plan.Nodes)
	depth := maxDepth(plan)
	risk := 0.5
	confidence := "low"
	complexity := float64(nodeCount)*0.1 + float64(depth)*0.2
	composite := math.Max(0.0, 1.0-risk-complexity*0.1)
	return PlanScore{
		NodeCount: nodeCount,
		Depth:     depth,
		Risk:      risk,
		Confidence: confidence,
		Complexity: math.Round(complexity*10000) / 10000,
		Composite: math.Round(composite*10000) / 10000,
	}
}

type Reflector struct {
	Events []map[string]any
}

func NewReflector(events []map[string]any) *Reflector {
	if events == nil {
		events = []map[string]any{}
	}
	return &Reflector{Events: events}
}

func (r *Reflector) Reflect(runID string) []Lesson {
	var lessons []Lesson
	lessons = append(lessons, r.detectFailurePatterns(runID)...)
	lessons = append(lessons, r.detectInefficiencies(runID)...)
	lessons = append(lessons, r.detectHandoffPatterns(runID)...)
	return lessons
}

func (r *Reflector) detectFailurePatterns(runID string) []Lesson {
	var lessons []Lesson
	taskFailures := map[string]int{}
	for _, e := range r.Events {
		if e["type"] == "task_status" && e["status"] == "[!]" {
			tid, _ := e["task_id"].(string)
			if tid != "" {
				taskFailures[tid]++
			}
		}
	}
	for tid, count := range taskFailures {
		if count >= 2 {
			lessons = append(lessons, Lesson{
				LessonID: fmt.Sprintf("lesson-%d", len(lessons)+1),
				RunID:    runID,
				Insight:  fmt.Sprintf("Task '%s' failed %d times; consider retry or fallback strategy.", tid, count),
				Severity: "warn",
				Tags:     []string{"failure", tid},
			})
		}
	}
	return lessons
}

func (r *Reflector) detectInefficiencies(runID string) []Lesson {
	var lessons []Lesson
	claimCounts := map[string]int{}
	for _, e := range r.Events {
		if e["type"] == "task_claim" {
			tid, _ := e["task_id"].(string)
			if tid != "" {
				claimCounts[tid]++
			}
		}
	}
	for tid, count := range claimCounts {
		if count >= 3 {
			lessons = append(lessons, Lesson{
				LessonID: fmt.Sprintf("lesson-%d", len(lessons)+1),
				RunID:    runID,
				Insight:  fmt.Sprintf("Task '%s' was claimed %d times; review ownership logic.", tid, count),
				Severity: "info",
				Tags:     []string{"efficiency", tid},
			})
		}
	}
	return lessons
}

func (r *Reflector) detectHandoffPatterns(runID string) []Lesson {
	handoffs := 0
	for _, e := range r.Events {
		if e["type"] == "human_handoff" || e["status"] == "[?]" {
			handoffs++
		}
	}
	if handoffs > 1 {
		return []Lesson{{
			LessonID: fmt.Sprintf("lesson-%d", 1),
			RunID:    runID,
			Insight:  fmt.Sprintf("Run had %d human handoffs; consider automating or simplifying decision gates.", handoffs),
			Severity: "warn",
			Tags:     []string{"handoff"},
		}}
	}
	return []Lesson{}
}

func (r *Reflector) ImprovePlan(plan *Plan, lessons []Lesson, constraints map[string]any) map[string]any {
	if plan == nil {
		return map[string]any{"plan": (*Plan)(nil), "proposals": []ImprovementProposal{}}
	}
	nodes := make([]*PlanNode, len(plan.Nodes))
	copy(nodes, plan.Nodes)
	seen := map[string]bool{}
	var proposals []ImprovementProposal
	for _, lesson := range lessons {
		contains := func(s []string, v string) bool {
			for _, x := range s {
				if x == v {
					return true
				}
			}
			return false
		}
		if contains(lesson.Tags, "failure") && contains(lesson.Tags, "failed") {
			target := extractTaskID(lesson.Insight)
			proposals = append(proposals, ImprovementProposal{
				ProposalID:   fmt.Sprintf("prop-%d", len(proposals)+1),
				LessonID:     lesson.LessonID,
				TargetNodeID: target,
				Action:       "add_dependency",
				Detail:       fmt.Sprintf("Add fallback or retry dependency for '%s' due to repeated failures.", target),
				Confidence:   0.75,
			})
		}
		if contains(lesson.Tags, "efficiency") && contains(lesson.Tags, "claimed") {
			target := extractTaskID(lesson.Insight)
			proposals = append(proposals, ImprovementProposal{
				ProposalID:   fmt.Sprintf("prop-%d", len(proposals)+1),
				LessonID:     lesson.LessonID,
				TargetNodeID: target,
				Action:       "reassign",
				Detail:       fmt.Sprintf("Reassign '%s' to a more stable owner.", target),
				Confidence:   0.6,
			})
		}
		if contains(lesson.Tags, "handoff") {
			proposals = append(proposals, ImprovementProposal{
				ProposalID: fmt.Sprintf("prop-%d", len(proposals)+1),
				LessonID:   lesson.LessonID,
				Action:     "add_node",
				Detail:     "Add automation gate to reduce human handoff frequency.",
				Confidence: 0.5,
			})
		}
	}
	maxNodes := 0
	if constraints != nil {
		if v, ok := constraints["max_nodes"].(int); ok {
			maxNodes = v
		}
	}
	for _, p := range proposals {
		if p.Action == "add_node" && !seen[p.ProposalID] {
			if maxNodes > 0 && len(nodes) >= maxNodes {
				continue
			}
			nodes = append(nodes, NewPlanNode(fmt.Sprintf("node-%s", p.ProposalID), "task", p.Detail, []string{}))
			seen[p.ProposalID] = true
		}
	}
	improved := NewPlan(plan.PlanID, plan.Goal, nodes, map[string]any{"improvements": collectActions(proposals)})
	for k, v := range plan.Metadata {
		improved.Metadata[k] = v
	}
	return map[string]any{
		"plan":      improved,
		"proposals": proposals,
	}
}

type CollabPlanner struct {
	estimator any
}

func NewCollabPlanner(estimator any) *CollabPlanner {
	return &CollabPlanner{estimator: estimator}
}

func (c *CollabPlanner) Build(goal string, constraints map[string]any) *Plan {
	if constraints == nil {
		constraints = map[string]any{}
	}
	decomposer := &GoalDecomposer{}
	plan := decomposer.Decompose(goal, constraints)
	if plan == nil {
		return nil
	}
	planner := &Planner{}
	ranked := planner.Rank([]*Plan{plan})
	if len(ranked) > 0 {
		plan = ranked[0].Plan
	}
	if c.estimator != nil {
		plan.Metadata["negotiation"] = "accepted"
	}
	return plan
}

func maxDepth(plan *Plan) int {
	if len(plan.Nodes) == 0 {
		return 0
	}
	depths := map[string]int{}
	for _, n := range plan.Nodes {
		depths[n.ID] = 1
	}
	for _, n := range plan.Nodes {
		for _, dep := range n.DependsOn {
			if d, ok := depths[dep]; ok {
				depths[n.ID] = max(depths[n.ID], d+1)
			}
		}
	}
	m := 0
	for _, d := range depths {
		if d > m {
			m = d
		}
	}
	return m
}

func extractTaskID(insight string) string {
	parts := strings.Split(insight, "'")
	if len(parts) >= 2 {
		return parts[1]
	}
	return "unknown"
}

func collectActions(proposals []ImprovementProposal) []string {
	actions := make([]string, 0, len(proposals))
	for _, p := range proposals {
		actions = append(actions, p.Action)
	}
	return actions
}

func sanitizeGoal(goal string) string {
	goal = strings.TrimSpace(goal)
	goal = replaceAll(goal, "[^a-z0-9_-]+", "-")
	return limitID(goal, 40)
}

func replaceAll(s, pattern, replacement string) string {
	// simple regex-free replacement for non-regexp build tags
	var result strings.Builder
	last := 0
	for i := 0; i < len(s); {
		if i+len(pattern) <= len(s) && s[i:i+len(pattern)] == pattern {
			result.WriteString(s[last:i])
			result.WriteString(replacement)
			last = i + len(pattern)
			i = last
		} else {
			i++
		}
	}
	result.WriteString(s[last:])
	return result.String()
}

func limitID(s string, max int) string {
	if len(s) > max {
		return s[:max]
	}
	if s == "" {
		return "plan"
	}
	return s
}

func extractVerbs(goal string) []string {
	words := strings.Fields(goal)
	var verbs []string
	for _, w := range words {
		clean := strings.Trim(w, ".,!?:;")
		if len(clean) > 2 && clean[0] >= 'A' && clean[0] <= 'Z' {
			verbs = append(verbs, clean)
		}
	}
	if len(verbs) == 0 {
		return []string{goal}
	}
	return verbs
}
